### PR TITLE
Fix range deletion tombstone ingestion with global seqno

### PR DESCRIPTION
**Pull request by jeffrey-xiao**
If we are writing a global seqno for an ingested file, the range tombstone metablock gets accessed and put into the cache during ingestion preparation. At the time, the global seqno of the ingested
file has not yet been determined, so the cached block will not have a global seqno. When the file is ingested and we read its range tombstone metablock, it will be returned from the cache with no global seqno. In that case, we use the actual seqnos stored in the range tombstones, which are all zero, so the tombstones cover nothing.

This commit removes global_seqno_ variable from Block. When iterating over a block, the global seqno for the block is determined by the iterator instead of storing this mutable attribute in Block. Additionally, this commit adds a regression test to check that keys are deleted when ingesting a file with a global seqno and range deletion tombstones.

### CODE REVIEW

**Comment by ajkr** [comment]
@siying what do you think about moving `global_seqno_` out of `Block` and instead obtaining it from `BlockBasedTableReader`? That prevents us from seeing wrong seqnum for blocks that are accessed and cached during file ingestion preparation. There are a couple benefits.

(1) we can remove this workaround for data blocks: https://github.com/facebook/rocksdb/blob/1729779b85e29f190612f82592e7070b6d24f58b/db/external_sst_file_ingestion_job.cc#L409-L414
(2) seqnums for meta-blocks like range tombstone meta-block will be correct regardless of whether they're placed in cache during file ingestion preparation

**Comment by siying** [comment]
@ajkr I think it is a good idea.

**Comment by ajkr** [comment]
I think this is an important fix. If Jeffrey's not available to rebase I can do it once review looks promising.

**Comment by ajkr** [comment]
@ltamasi Rebased anyways as I still think this is an important fix. Please let me know any questions.

File: db/external_sst_file_basic_test.cc [hunk 1]
```diff
@@ -796,6 +796,45 @@ TEST_F(ExternalSSTFileBasicTest, VerifyChecksumReadahead) {
...
+  Reopen(options);
+  SstFileWriter sst_file_writer(EnvOptions(), options);
+
+  // file.sst (delete 0 => 30)
+  std::string file = sst_files_dir_ + "file.sst";
+  ASSERT_OK(sst_file_writer.Open(file));
+  ASSERT_OK(sst_file_writer.DeleteRange(Key(0), Key(30)));
+  ExternalSstFileInfo file_info;
+  Status s = sst_file_writer.Finish(&file_info);
+  ASSERT_TRUE(s.ok()) << s.ToString();
```
Comment (ltamasi) [inline db/external_sst_file_basic_test.cc]: You can just say `ASSERT_OK(s);` here, or even combine it with the previous line (`ASSERT_OK(sst_file_writer.Finish(&file_info));`).

File: db/external_sst_file_basic_test.cc [hunk 2]
```diff
@@ -796,6 +796,45 @@ TEST_F(ExternalSSTFileBasicTest, VerifyChecksumReadahead) {
...
+  ifo.move_files = true;
+  ifo.snapshot_consistency = true;
+  ifo.allow_global_seqno = true;
+  ifo.write_global_seqno = true;
+  ifo.verify_checksums_before_ingest = false;
+  s = db_->IngestExternalFile({file}, ifo);
+
+  ASSERT_OK(s);
+  for (int i = 5; i < 25; i++) {
+    ASSERT_EQ(Get(Key(i)), "NOT_FOUND");
```
Comment (ltamasi) [inline db/external_sst_file_basic_test.cc]: `ASSERT_TRUE(Get(Key(i)).IsNotFound())` would be a bit cleaner here.

File: table/block_based/block.h [hunk 3]
```diff
@@ -150,8 +150,7 @@ class BlockReadAmpBitmap {
 class Block {
  public:
   // Initialize the block with the specified contents.
-  explicit Block(BlockContents&& contents, SequenceNumber _global_seqno,
-                 size_t read_amp_bytes_per_bit = 0,
+  explicit Block(BlockContents&& contents, size_t read_amp_bytes_per_bit = 0,
```
Comment (ltamasi) [inline table/block_based/block.h]: The problem here (not with the PR per se but with the pre-existing default parameter values) is that removing `_global_seqno` changes the meaning of code like this:

```
Block reader1(std::move(contents), kDisableGlobalSequenceNumber); // kDisableGlobalSequenceNumber is now read_amp_bytes_per_bit !!!
```

We should review all places where `Block`s are constructed and fix any problems like this.

**Comment by ltamasi** [comment]
Also, could you mention the fix in HISTORY.md?

**Comment by ajkr** [comment]
I need to open a new PR to push changes to the branch. Doing that and replying to comments there: https://github.com/facebook/rocksdb/pull/6429.

**Pull request closed without merging**